src/framework/standard/check.rs

```diff
@@ -0,0 +1,221 @@
+    /// [`Reason::Log`]: struct.Reason.html#variant.Log
+    pub fn new_log<D>(d: D) -> Self
+        where D: fmt::Display {
+        CheckResult::Failure(Reason::Log(d.to_string()))
+    }
+
+    /// Creates a new [`CheckResult::Failure`] with [`Reason::Unknown`].
+    ///
+    /// [`CheckResult::Failure`]: enum.CheckResult.html#variant.Failure
+    /// [`Reason::Unknown`]: struct.Reason.html#variant.Unknown
+    pub fn new_unknown() -> Self {
+        CheckResult::Failure(Reason::Unknown)
+    }
+
+    /// Creates a new [`CheckResult::Failure`] with [`Reason::UserAndLog`].
+    ///
+    /// [`CheckResult::Failure`]: enum.CheckResult.html#variant.Failure
+    /// [`Reason::UserAndLog`]: struct.Reason.html#variant.UserAndLog
+    pub fn new_user_and_log<D>(user: D, log: D) -> Self
+        where D: fmt::Display {
+        CheckResult::Failure(Reason::UserAndLog {
+            user: user.to_string(),
+            log: log.to_string(),
+        })
+    }
+
+    /// Returns `true` if [`CheckResult`] is [`CheckResult::Success`] and
+    /// `false` if not.
+    ///
+    /// [`CheckResult`]: enum.CheckResult.html
+    /// [`CheckResult::Success`]: enum.CheckResult.html#variant.Success
+    pub fn is_success(&self) -> bool {
+        if let CheckResult::Success = self {
+            return true;
+        }
+
+        false
+    }
+}
+
+impl From<bool> for CheckResult {
+    fn from(succeeded: bool) -> Self {
+        if succeeded {
+            CheckResult::Success
+        } else {
+            CheckResult::Failure(Reason::Unknown)
+        }
+    }
+}
+
+impl From<Reason> for CheckResult {
+    fn from(reason: Reason) -> Self {
+        CheckResult::Failure(reason)
+    }
+}
+
+/// A check can be part of a command or group and will be executed to
+/// determine whether a user is permitted to use related item.
+///
+/// Additionally, a check may hold additional settings.
+pub struct Check {
+    /// Name listed in help-system.
+    pub name: String,
+    /// Function that will be executed.
+    pub function: Box<CheckFunction>,
+    /// Whether a check should be evaluated in the help-system.
+    /// `false` will ignore check and won't fail execution.
+    pub check_in_help: bool,
+    /// Whether a check shall be listed in the help-system.
+    /// `false` won't affect whether the check will be evaluated help,
+    /// solely `check_in_help` sets this.
+    pub display_in_help: bool,
+}
+
+impl Check {
+    pub(crate) fn new<F>(name: &str, function: F, check_in_help: bool, display_in_help: bool) -> Self
+    where F: Fn(&mut Context, &Message, &mut Args, &CommandOptions) -> CheckResult  + Send + Sync + 'static {
+        Self {
+            name: name.to_string(),
+            function: Box::new(function),
+            check_in_help,
+            display_in_help,
+        }
+    }
+}
+
+impl Debug for Check {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Check")
```

Comment:
This redundantly allocates strings on every debug call. The code also looks ugly.
```rust
f.debug_struct("Check")
   .field("name", &self.name)
   .field("function", &"<fn>")
   .field("check_in_help", &self.check_in_help)
   .field("display_in_help", &self.display_in_help)
   .finish()
```